docs/exernal-services.md

```diff
@@ -0,0 +1,206 @@
+
+## AuthoringSite model
+Additionally there is an AuthoringSite model is how a convenience 'create XYZ' button can be added to `/authoring` page.
+
+# Teacher and Student Reports or Dashboards
+As described above ExternalReports are used to provide reports and dashboards for ExternalActivities. The ExternalReport model is also used for Researcher reports.
+
+The ExternalReport has a Client which indicates what domain the report is hosted on, so the portal can validate its requests.
+
+## Types
+An ExternalReport has 2 types for teachers:
+- offering
+- class
+
+It can also be marked as "allowed for students"
+
+The other types of ExternalReports are described in the Researcher section
+
+### Offering Report
+An offering report is associated with an assignment in the Portal. So in places where teachers see the assignment they also see a button to run this report.
+
+The report has to be associated with the assigned ExternalActivity either directly or indirectly through the source_type.
+
+### Class Report
+A class report is for classes. In places where teachers see classes they will see a button to run this report.
+
+The report has to be associated directly with the ExternalActivity. The indirect use of source_type is not supported for Class reports.
+
+When a class has 1 or more assignments with a Class Report then the report button is shown at the class level of the UI.
+
+### Student Report
+If the report is allowed for students, and the report is an offering report, then the student will see a report button next to the assignment.
+
+When the student report is launched, the user id of the current user is passed as a `studentId` to the report. This is in addition to the other parameters below. Reference: https://github.com/concord-consortium/rigse/blob/d81d5a1da69253418abde7e0aa7baac4a1831a66/rails/app/models/external_report.rb#L39
+Also the learner (student, offering) is added to the access grant of the token.
+
+## Launching
+The offering_controller provides the following actions for reports:
+- student_report
+- report
+- external_report
+
+The Portal adds several parameters onto the URL provided by the ExternalReport. They can be seen here: https://github.com/concord-consortium/rigse/blob/d81d5a1da69253418abde7e0aa7baac4a1831a66/rails/app/models/external_report.rb#L68
+
+## Authorization
+Like ExternalActivities, ExternalReports provide a token the report can use to call back to the Portal. Unlike ExternalActivities this token has a client. So when an ExternalReport is setup in the Portal a Client is needed and must have the correct settings. Our teacher reports are normally SPAs, so the Client should have a type of public, and it should specify the domain of the SPA.
+
+If the SPA report is also going to support OAuth2, so it can be run without being launched from the portal with a token, then the redirect url must be specified. This redirect url is used by the Portal after the user is signed in.
```

Comment:
```suggestion
If the SPA report is also going to support OAuth2 (so it can be run without being launched from the portal with a token), then the redirect url must be specified. This redirect url is used by the Portal after the user is signed in.
```